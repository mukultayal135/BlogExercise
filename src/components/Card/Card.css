.card {
  height: 450px;
  width: 300px;
  display: flex;
  flex-direction: column;
  border: 2px solid lightgray;
}
