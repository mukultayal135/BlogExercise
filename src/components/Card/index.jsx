import PropTypes from 'prop-types';
import { useState } from 'react';
import Img from '../Img';
import PostContent from '../PostContent';
import Reactions from '../Reactions';
import './Card.css';
import { UPDATE_BLOG_DATA } from '../../constants/apiEndPoints';
import { getFormattedDateFromUtcDate } from '../../utils/common/date';
import makeRequest from '../../utils/makeRequest';

const Card = ({ blogData }) => {
  const [claps, setClaps] = useState(blogData.claps);
  const [like, setLike] = useState(blogData.liked);
  const onReact = async () => {
    try {
      await makeRequest(UPDATE_BLOG_DATA(blogData.id), () => {}, {
        data: { claps: claps + 1 },
      });
      setClaps(claps + 1);
    } catch (e) {
      //
    }
  };
  const onLike = async () => {
    try {
      await makeRequest(UPDATE_BLOG_DATA(blogData.id), () => {}, {
        data: { liked: !like },
      });
      setLike(!like);
    } catch (e) {
      //
    }
  };
  return (
    <div className="card" data-testid="card">
      <Img path={blogData.image} />
      <PostContent
        title={blogData.title}
        description={blogData.description}
        readingTime={blogData.reading_time}
        date={getFormattedDateFromUtcDate(blogData.date)}
      />
      <Reactions onReact={onReact} count={claps} onLike={onLike} liked={like} />
    </div>
  );
};
Card.propTypes = {
  blogData: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    reading_time: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
    claps: PropTypes.number.isRequired,
    image: PropTypes.string.isRequired,
    liked: PropTypes.bool.isRequired,
  }).isRequired,
};
export default Card;
