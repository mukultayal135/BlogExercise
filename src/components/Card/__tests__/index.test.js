/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { fireEvent, screen, render, waitFor } from '@testing-library/react';
import Card from '..';
import makeRequest from '../../../utils/makeRequest';

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
}));
jest.mock('../../../utils/makeRequest', () => jest.fn());
describe('Card', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  const mockProps = {
    id: 2,
    date: '6/29/2022 4:52:48 PM UTC',
    reading_time: '4 mins',
    title: 'mock title 2',
    description: 'mock description 2',
    claps: 12,
    liked: false,
    image: 'abstract.png',
  };
  it('should render Card component', () => {
    const { asFragment } = render(<Card blogData={mockProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should increase clap count by 1 when click on clap icon', async () => {
    makeRequest.mockResolvedValue({ claps: 12 });
    render(<Card blogData={mockProps} />);
    const clapCount = screen.getByTestId('count-clap');
    const initialCount = clapCount.textContent;
    expect(initialCount).toBe('12');
    const clapIcon = screen.getByAltText('clapImage');
    fireEvent.click(clapIcon);
    await waitFor(() => {
      expect(clapCount.textContent).toBe('13');
    });
  });
  it('should change heart icon to red when click on black heart icon', async () => {
    makeRequest.mockResolvedValue({ like: true });
    render(<Card blogData={mockProps} />);
    const likeIcon = screen.getByAltText('heartImg');
    expect(likeIcon.src).toContain('heart-black.svg');
    fireEvent.click(likeIcon);
    await waitFor(() => {
      expect(likeIcon.src).toContain('heart-red.svg');
    });
  });
  it('should change heart icon to black when click on red heart icon', async () => {
    makeRequest.mockResolvedValue({ like: false });
    render(<Card blogData={{ ...mockProps, liked: true }} />);
    const likeIcon = screen.getByAltText('heartImg');
    expect(likeIcon.src).toContain('heart-red.svg');
    fireEvent.click(likeIcon);
    await waitFor(() => {
      expect(likeIcon.src).toContain('heart-black.svg');
    });
  });
  // it('should change get error when clap not updated', async () => {
  //   makeRequest.mockRejectedValue();
  //   render(<Card blogData={{ ...mockProps, liked: true }} />);
  //   const likeIcon = screen.getByAltText('heartImg');
  //   expect(likeIcon.src).toContain('heart-red.svg');
  //   fireEvent.click(likeIcon);
  //   await waitFor(() => {
  //     expect(likeIcon.src).toContain('heart-red.svg');
  //   });
  // });
});
