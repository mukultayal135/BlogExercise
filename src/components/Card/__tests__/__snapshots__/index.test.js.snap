// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render Card component 1`] = `
<DocumentFragment>
  <div
    class="card"
    data-testid="card"
  >
    <div>
      <img
        alt=""
        class="card-image"
        src="abstract.png"
      />
    </div>
    <div
      class="card-content card-padding"
    >
      <div
        class="date-time"
      >
        <div
          class="date"
        >
          29th
      June, 
      2022
        </div>
        <div
          class="read-time"
        >
          4 mins
        </div>
      </div>
      <div
        class="card-heading"
      >
        mock title 2
      </div>
      <div
        class="card-description"
      >
        mock description 2
      </div>
    </div>
    <hr />
    <div
      class="reactions"
    >
      <div
        class="clap"
      >
        <img
          alt="clapImage"
          src="clapping.svg"
        />
        <div
          class="count"
          data-testid="count-clap"
        >
          12
        </div>
      </div>
      <div
        class="heart"
      >
        <img
          alt="heartImg"
          src="heart-black.svg"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
